# Similar products

In [1]:
% matplotlib inline
import pickle

import numpy as np
import pandas as pd
from sklearn.neighbors import NearestNeighbors


## Load video games review dataframe

In [2]:
with open('pickle/video_games.pkl', 'rb') as input:
    video_games = pickle.load(input)

In [3]:
video_games.head()

overall  \
asin       reviewerID                
0700099867 A2HD75EMZR8QLN        1   
           A3UR8NLLY1ZHCX        4   
           A1INA0F5CWW3J4        1   
           A1DLMTOTHQ4AST        3   
           A361M14PU2GUEG        4   

                                                                  reviewText  \
asin       reviewerID                                                          
0700099867 A2HD75EMZR8QLN  Installing the game was a struggle (because of...   
           A3UR8NLLY1ZHCX  If you like rally cars get this game you will ...   
           A1INA0F5CWW3J4  1st shipment received a book instead of the ga...   
           A1DLMTOTHQ4AST  I got this version instead of the PS3 version,...   
           A361M14PU2GUEG  I had Dirt 2 on Xbox 360 and it was an okay ga...   

                                                 reviewerName  \
asin       reviewerID                                           
0700099867 A2HD75EMZR8QLN                                 123   
           A3UR8NLLY1ZHCX  Alejandro Henao "Electronic Junky"   
           A1INA0F5CWW3J4          Amazon Shopper "Mr.Repsol"   
           A1DLMTOTHQ4AST                            ampgreen   
           A361M14PU2GUEG        Angry Ryan "Ryan A. Forrest"   

                                                                   summary  \
asin       reviewerID                                                        
0700099867 A2HD75EMZR8QLN         Pay to unlock content? I don't think so.   
           A3UR8NLLY1ZHCX                                  Good rally game   
           A1INA0F5CWW3J4                                        Wrong key   
           A1DLMTOTHQ4AST  awesome game, if it did not crash frequently !!   
           A361M14PU2GUEG                                           DIRT 3   

                                date  helpfulRate  helpfulCount  rank  \
asin       reviewerID                                                   
0700099867 A2HD75EMZR8QLN 2012-07-09     0.666667            12  6629   
           A3UR8NLLY1ZHCX 2013-06-30    -1.000000             0  6629   
           A1INA0F5CWW3J4 2014-06-28    -1.000000             0  6629   
           A1DLMTOTHQ4AST 2011-09-14     0.700000            10  6629   
           A361M14PU2GUEG 2011-06-14     1.000000             2  6629   

                                                                 description  \
asin       reviewerID                                                          
0700099867 A2HD75EMZR8QLN  Dirt 3 is a popular rally racing game for Play...   
           A3UR8NLLY1ZHCX  Dirt 3 is a popular rally racing game for Play...   
           A1INA0F5CWW3J4  Dirt 3 is a popular rally racing game for Play...   
           A1DLMTOTHQ4AST  Dirt 3 is a popular rally racing game for Play...   
           A361M14PU2GUEG  Dirt 3 is a popular rally racing game for Play...   

                            price title brand  reviewLength  summaryLength  \
asin       reviewerID                                                        
0700099867 A2HD75EMZR8QLN  246.63                       779             40   
           A3UR8NLLY1ZHCX  246.63                       304             15   
           A1INA0F5CWW3J4  246.63                       282              9   
           A1DLMTOTHQ4AST  246.63                      3489             47   
           A361M14PU2GUEG  246.63                       294              6   

                           reviewerNameLength  
asin       reviewerID                          
0700099867 A2HD75EMZR8QLN                 3.0  
           A3UR8NLLY1ZHCX                34.0  
           A1INA0F5CWW3J4                26.0  
           A1DLMTOTHQ4AST                 8.0  
           A361M14PU2GUEG                28.0

## Load products metadata

In [4]:
video_games_meta = pd.read_json('./data/video-games.meta.parsed.json')
video_games_meta.head()

asin brand                        categories  \
0  0078764343   NaN  [[Video Games, Xbox 360, Games]]   
1  043933702X   NaN        [[Video Games, PC, Games]]   
2  0439339987   NaN        [[Video Games, PC, Games]]   
3  0439342260   NaN        [[Video Games, PC, Games]]   
4  0439339960   NaN        [[Video Games, PC, Games]]   

                                         description  \
0                                  Brand new sealed!   
1  In Stock NOW.  Eligible for FREE Super Saving ...   
2  Grandma Groupers kelp seeds are missing and wi...   
3  This software is BRAND NEW. Packaging may diff...   
4               a scholastic clubs fairs cd rom game   

                                               imUrl  price  \
0  http://ecx.images-amazon.com/images/I/513h6dPb...  37.98   
1  http://ecx.images-amazon.com/images/I/61KKRndV...  23.50   
2  http://ecx.images-amazon.com/images/I/416QZg89...   8.95   
3  http://ecx.images-amazon.com/images/I/61Wvu-Uj...    NaN   
4  http://ecx.images-amazon.com/images/I/51k3oRCF...    NaN   

                                             related               salesRank  \
0  {'also_bought': ['B000TI836G', 'B003Q53VZC', '...  {'Video Games': 28655}   
1  {'also_viewed': ['B000067NP1', '0439573947', '...  {'Video Games': 44080}   
2  {'also_bought': ['B000314VVU', 'B000PXUOTE', '...  {'Video Games': 49836}   
3                    {'also_viewed': ['043934302X']}  {'Video Games': 49156}   
4                    {'also_viewed': ['B00028D7TG']}  {'Video Games': 52262}   

  title  
0   NaN  
1   NaN  
2   NaN  
3   NaN  
4   NaN

### Playstation
We try to find similar product to the playstation aka others console. We can first try to use the related column

In [5]:
playstation = video_games_meta[video_games_meta.asin == 'B00BGA9WK2']
playstation

asin brand                                         categories  \
45933  B00BGA9WK2   NaN  [[Electronics, Television & Video, Blu-ray Pla...   

      description                                              imUrl  price  \
45933              http://ecx.images-amazon.com/images/I/41omR-LT...  28.12   

                                                 related            salesRank  \
45933  {'also_bought': ['B00BGA9X9W', 'B00CXCCI8A', '...  {'Video Games': 13}   

      title  
45933   NaN

#### Related column
##### Buy after viewing

In [6]:
buy_after_viewing_asin = playstation.related.values[0]['buy_after_viewing']
buy_after_viewing = video_games_meta[video_games_meta.asin.isin(buy_after_viewing_asin)]
buy_after_viewing

asin brand                                         categories  \
45932  B00BGA9X9W   NaN  [[Video Games, PlayStation 4, Accessories, Con...   
47181  B00CXCCH1I   NaN              [[Video Games, PlayStation 4, Games]]   

                                             description  \
45932                                                      
47181  With its first new major sports league partner...   

                                                   imUrl  price  \
45932  http://ecx.images-amazon.com/images/I/41ih-c4l...  59.99   
47181  http://ecx.images-amazon.com/images/I/510xZgGR...  59.99   

                                                 related  \
45932  {'also_bought': ['B00BGA9WK2', 'B00CXCCI8A', '...   
47181  {'also_bought': ['B00BI83EVU', 'B00BGA9X9W', '...   

                  salesRank title  
45932   {'Video Games': 15}   NaN  
47181  {'Video Games': 102}   NaN

##### Also bought

In [7]:
also_bought_asin = playstation.related.values[0]['also_bought']
also_bought = video_games_meta[video_games_meta.asin.isin(also_bought_asin)]
also_bought.head()

asin brand                                         categories  \
33598  B003NSLGW2   NaN  [[Video Games, More Systems, PlayStation 2, Ha...   
44201  B009GZX49Y   NaN              [[Video Games, PlayStation 3, Games]]   
45932  B00BGA9X9W   NaN  [[Video Games, PlayStation 4, Accessories, Con...   
45934  B00BGAA3S2   NaN  [[Toys & Games, Electronics for Kids], [Video ...   
45935  B00BGAA29M   NaN  [[Video Games, Kids & Family, PlayStation 4, G...   

                                             description  \
33598  Grip- iT is the newest gaming accessory that e...   
44201  Experience the series that redefined the actio...   
45932                                                      
45934                                                      
45935                                                      

                                                   imUrl  price  \
33598  http://ecx.images-amazon.com/images/I/41QPWHza...   4.99   
44201  http://ecx.images-amazon.com/images/I/61IJHp-q...  29.99   
45932  http://ecx.images-amazon.com/images/I/41ih-c4l...  59.99   
45934  http://ecx.images-amazon.com/images/I/31vC6fcF...  59.99   
45935  http://ecx.images-amazon.com/images/I/51BqJurZ...  59.99   

                                                 related  \
33598  {'also_bought': ['B00GME9XWE', 'B00HHUSNPU', '...   
44201  {'also_bought': ['B000P46NMA', 'B0050SXFKE', '...   
45932  {'also_bought': ['B00BGA9WK2', 'B00CXCCI8A', '...   
45934  {'also_bought': ['B00D8S4GRY', 'B00J3Y56WG', '...   
45935  {'also_bought': ['B00DUARBTA', 'B00H5V9S6E', '...   

                  salesRank title  
33598   {'Video Games': 57}   NaN  
44201  {'Video Games': 681}   NaN  
45932   {'Video Games': 15}   NaN  
45934   {'Video Games': 46}   NaN  
45935  {'Video Games': 478}   NaN

##### Bought together

In [8]:
bought_together_asin = playstation.related.values[0]['bought_together']
bought_together = video_games_meta[video_games_meta.asin.isin(bought_together_asin)]
bought_together.head()

asin brand                                         categories  \
45932  B00BGA9X9W   NaN  [[Video Games, PlayStation 4, Accessories, Con...   
47179  B00CXCCI8A   NaN  [[Video Games, Kids & Family, PlayStation 4, G...   

                                             description  \
45932                                                      
47179  Step onto the field and immerse yourself in th...   

                                                   imUrl  price  \
45932  http://ecx.images-amazon.com/images/I/41ih-c4l...  59.99   
47179  http://ecx.images-amazon.com/images/I/51yQq3Dn...  59.99   

                                                 related            salesRank  \
45932  {'also_bought': ['B00BGA9WK2', 'B00CXCCI8A', '...  {'Video Games': 15}   
47179  {'also_bought': ['B00BGA9X9W', 'B00BGA9WK2', '...  {'Video Games': 51}   

      title  
45932   NaN  
47179   NaN

Seems like most of the related products are games for Playstation or accessories.

#### NearestNeighbors
We can try to use a NearestNeighbors method to find similar product. However, there is only the price column that is a number.  

##### Categories
We transform the categories to a one hot encoding format.  
An Amazon categories is represented by a list of categories, sub categories, subsub categories ...

For example `['Electronics', 'Television & Video', 'Blu-ray Players & Recorders']` or `['Video Games', 'PlayStation 4', 'Consoles']`. The leftest category is the most specific.

We can encode a list `['Electronics', 'Television & Video', 'Blu-ray Players & Recorders']` as a category, but we can also consider the subcategory as a category to. So we will have `'Electronics'`, `'Television & Video'`and `'Blu-ray Players & Recorders'`. We will use this representation as it allows more combinations and makes more easier the neighbor search. 

For example `CategoryA -> SubCategoryZ -> SubsubCategoryA` is very similar to `CategoryA -> SubCategoryY -> SubsubCategoryA`. Only the `Subcategory` makes it different.

In [9]:
flat_list_categories = [item for sublist in video_games_meta.categories.values for item in sublist] # Flatten the list of categories
flat_list_categories = [item for sublist in flat_list_categories for item in sublist] # Flatten the categories, subcategories etc..
flat_list_categories = list(set(flat_list_categories)) # Remove duplicates
flat_list_categories[:10]

['Legal',
 'Home Office Furniture',
 'Board Games',
 'Personal Finance',
 'Hand Tools',
 'Waterskiing & Towsports',
 'R&B',
 'Database',
 'Gaming Mice',
 'TV Remote Controls']

In [10]:
print("We get", len(flat_list_categories), "number of categories")

We get 468 number of categories


In [11]:
from sklearn.preprocessing import OneHotEncoder
from sklearn import preprocessing
label_encoder = preprocessing.LabelEncoder()

# Encode category string to number
label_encoder.fit(flat_list_categories)
categories = label_encoder.transform(flat_list_categories)

# Encode to one hot
on_hot_encoder = OneHotEncoder()
on_hot_encoder.fit([[cat] for cat in categories])

OneHotEncoder(categorical_features='all', dtype=<class 'numpy.float64'>,
       handle_unknown='error', n_values='auto', sparse=True)

In [12]:
# The one hot encoding for the 223 category
on_hot = np.asarray(on_hot_encoder.transform(223).todense()).reshape(-1)
on_hot

array([ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
        0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
        0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
        0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
        0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
        0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
        0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
        0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
        0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
        0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
        0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
        0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
        0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
        0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0

In [13]:
def get_features_from_product_metadata(product):
    categories_labels = []
    for category_list in product.categories:
        categories_labels = np.concatenate([categories_labels, label_encoder.transform(category_list)])
    
    categories_one_hot = np.zeros(468) # 468 categories
    for cat_label in categories_labels:
        categories_one_hot = categories_one_hot + on_hot_encoder.transform(cat_label)
        
    features = list(np.asarray(categories_one_hot[0])[0]) # Categories features
    features.append(product.price) # Add price feature
    return features

# Create a list of products with onehot category and price features

video_games_meta_dropped_price_nan = video_games_meta[np.isfinite(video_games_meta['price'])]

X = []
X_asin = []
for i, product in video_games_meta_dropped_price_nan.iterrows():
    features = get_features_from_product_metadata(product)
    
    X.append(features)
    X_asin.append(product.asin)
    
    if i % 1000 == 0:
        print(i,"/",len(video_games_meta_dropped_price_nan))

0 / 43905
4000 / 43905
5000 / 43905
6000 / 43905
7000 / 43905
8000 / 43905
10000 / 43905
11000 / 43905
12000 / 43905
14000 / 43905
16000 / 43905
18000 / 43905
20000 / 43905
22000 / 43905
23000 / 43905
24000 / 43905
25000 / 43905
26000 / 43905
28000 / 43905
29000 / 43905
30000 / 43905
31000 / 43905
32000 / 43905
33000 / 43905
34000 / 43905
35000 / 43905
36000 / 43905
37000 / 43905
39000 / 43905
40000 / 43905
41000 / 43905
42000 / 43905
43000 / 43905
44000 / 43905
45000 / 43905
46000 / 43905
47000 / 43905
48000 / 43905
49000 / 43905
50000 / 43905


We now can use these products representation to find the nearest neighbors for a given product

In [19]:
# Return the indices of neighbors products
def get_nearest_neighbors_product(product, X, X_asin, n_neighbors=2):
    nbrs = NearestNeighbors(n_neighbors=n_neighbors, algorithm='ball_tree').fit(X)
    distances, indices = nbrs.kneighbors([get_features_from_product_metadata(product.squeeze())])
    
    asins = [X_asin[i] for i in indices[0]]
    return asins

In [20]:
asins = get_nearest_neighbors_product(playstation, X, X_asin)

In [21]:
video_games_meta[video_games_meta.asin == asins[1]]

asin brand       categories  \
14094  B000930DOG   NaN  [[Video Games]]   

                                             description  \
14094  The year is 1963. Your name is Bond, James Bon...   

                                                   imUrl  price  \
14094  http://ecx.images-amazon.com/images/I/515YPQDX...  27.95   

                                                 related  \
14094  {'also_bought': ['B00009X3V2', 'B00005V3EX', '...   

                    salesRank title  
14094  {'Video Games': 22785}   NaN

The similar product obtained for the playstation console is a James bond game. We would expect to get a similar console. This is due to the playstation price of $28.12, which is probably a mistake.

We try it on a Xbox game :

In [23]:
xbox_game = video_games_meta.iloc[0]

In [25]:
asins = get_nearest_neighbors_product(xbox_game, X, X_asin)
video_games_meta[video_games_meta.asin.isin(asins)]

asin brand                        categories  \
0      0078764343   NaN  [[Video Games, Xbox 360, Games]]   
24292  B00175GC4U   NaN  [[Video Games, Xbox 360, Games]]   

                                             description  \
0                                      Brand new sealed!   
24292  The Number One fighting game franchise in the ...   

                                                   imUrl  price  \
0      http://ecx.images-amazon.com/images/I/513h6dPb...  37.98   
24292  http://ecx.images-amazon.com/images/I/51yZJeFK...  37.99   

                                                 related  \
0      {'also_bought': ['B000TI836G', 'B003Q53VZC', '...   
24292  {'also_bought': ['B000FW8FK0', 'B002E1SBVS', '...   

                    salesRank title  
0      {'Video Games': 28655}   NaN  
24292  {'Video Games': 12736}   NaN

We get a similar product, which stay in the same category "Xbox360 games".

#### Get enough products to reach 10k reviews
We start from one xbox game and get 1000 nearest neighbors

In [33]:
asins = get_nearest_neighbors_product(xbox_game, X, X_asin, 1000)
video_games_meta[video_games_meta.asin.isin(asins)].head()

asin brand                              categories  \
0    0078764343   NaN        [[Video Games, Xbox 360, Games]]   
19   0545076412   NaN              [[Video Games, PC, Games]]   
24   0700026657   NaN              [[Video Games, PC, Games]]   
142  9882107753   NaN  [[Video Games, Xbox 360, Accessories]]   
152  9944860212   NaN              [[Video Games, PC, Games]]   

                                           description  \
0                                    Brand new sealed!   
19   PROTECT, BEAUTIFY, BUILD, GROW, LIVE, AND ENTE...   
24   Anno 2070, the newest version of the award-win...   
142  Professional and exclusively design for your X...   
152  This software is BRAND NEW. Packaging may diff...   

                                                 imUrl  price  \
0    http://ecx.images-amazon.com/images/I/513h6dPb...  37.98   
19   http://ecx.images-amazon.com/images/I/41CQQPbp...  39.99   
24   http://ecx.images-amazon.com/images/I/51R06L8c...  39.99   
142  http://ecx.images-amazon.com/images/I/4119cGkv...  35.99   
152  http://ecx.images-amazon.com/images/I/51P3ar3k...  39.99   

                                               related  \
0    {'also_bought': ['B000TI836G', 'B003Q53VZC', '...   
19   {'also_viewed': ['B000FHE6EY'], 'buy_after_vie...   
24   {'also_bought': ['B005OM6CGU', 'B001AWMCMQ', '...   
142  {'also_viewed': ['B00GJDJP3K'], 'buy_after_vie...   
152  {'also_viewed': ['B0007II11K', 'B00006C2HA', '...   

                  salesRank title  
0    {'Video Games': 28655}   NaN  
19   {'Video Games': 55546}   NaN  
24    {'Video Games': 7621}   NaN  
142  {'Video Games': 88073}   NaN  
152  {'Video Games': 53215}   NaN

Then we concatenate the reviews of the 1000 xbox video games

In [39]:
xbox_games_reviews = video_games.iloc[video_games.index.get_level_values('asin').isin(asins)]
print("Reviews count:", len(xbox_games_reviews))
xbox_games_reviews.head()

Reviews count: 10610


overall  \
asin       reviewerID                
B000006RGQ A3DABGGV3YAE2V        4   
           A1LBGXE29GG1IR        3   
           A11LS12ZU93SV6        2   
           ASHK9LB9MU5ZI         5   
           A2VRO53CM5EP28        5   

                                                                  reviewText  \
asin       reviewerID                                                          
B000006RGQ A3DABGGV3YAE2V  i would recommend getting this game for many r...   
           A1LBGXE29GG1IR  Throughout the years there have been dozens of...   
           A11LS12ZU93SV6  You know you're in trouble when other terrible...   
           ASHK9LB9MU5ZI   This game is very fun. I have only played a de...   
           A2VRO53CM5EP28  I played this game when i was a kid and forgot...   

                                                       reviewerName  \
asin       reviewerID                                                 
B000006RGQ A3DABGGV3YAE2V                                  C. Glass   
           A1LBGXE29GG1IR  Coffee Addicted Writer's Reviews "Billy"   
           A11LS12ZU93SV6                     DeRosset Myers "DeRo"   
           ASHK9LB9MU5ZI                                    "isams"   
           A2VRO53CM5EP28                                     ItsMe   

                                  summary       date  helpfulRate  \
asin       reviewerID                                               
B000006RGQ A3DABGGV3YAE2V  FIGHTING FORCE 2011-06-02         -1.0   
           A1LBGXE29GG1IR  Fighting Force 2012-10-14          1.0   
           A11LS12ZU93SV6    Pretty awful 2006-04-13          0.6   
           ASHK9LB9MU5ZI    Fighting fun! 2001-09-04         -1.0   
           A2VRO53CM5EP28        Fun game 2013-01-13         -1.0   

                           helpfulCount   rank  \
asin       reviewerID                            
B000006RGQ A3DABGGV3YAE2V             0  18320   
           A1LBGXE29GG1IR             1  18320   
           A11LS12ZU93SV6             5  18320   
           ASHK9LB9MU5ZI              0  18320   
           A2VRO53CM5EP28             0  18320   

                                                                 description  \
asin       reviewerID                                                          
B000006RGQ A3DABGGV3YAE2V  Fighting Force, from the same developers that ...   
           A1LBGXE29GG1IR  Fighting Force, from the same developers that ...   
           A11LS12ZU93SV6  Fighting Force, from the same developers that ...   
           ASHK9LB9MU5ZI   Fighting Force, from the same developers that ...   
           A2VRO53CM5EP28  Fighting Force, from the same developers that ...   

                           price title brand  reviewLength  summaryLength  \
asin       reviewerID                                                       
B000006RGQ A3DABGGV3YAE2V  38.24                       308             14   
           A1LBGXE29GG1IR  38.24                       208             14   
           A11LS12ZU93SV6  38.24                      1582             12   
           ASHK9LB9MU5ZI   38.24                       712             13   
           A2VRO53CM5EP28  38.24                       129              8   

                           reviewerNameLength  
asin       reviewerID                          
B000006RGQ A3DABGGV3YAE2V                 8.0  
           A1LBGXE29GG1IR                40.0  
           A11LS12ZU93SV6                21.0  
           ASHK9LB9MU5ZI                  7.0  
           A2VRO53CM5EP28                 5.0